applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -0,0 +1,856 @@
+        AssignLeftHandSideWakeNode(rLeftHandSideMatrix, lhs_total, data, row);
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::AssignLeftHandSideWakeNode(
+    MatrixType& rLeftHandSideMatrix,
+    const BoundedMatrix<double, NumNodes, NumNodes>& lhs_total,
+    const ElementalData<NumNodes, Dim>& data,
+    unsigned int& row) const
+{
+    // Filling the diagonal blocks (i.e. decoupling upper and lower dofs)
+    for (unsigned int column = 0; column < NumNodes; ++column)
+    {
+        rLeftHandSideMatrix(row, column) = lhs_total(row, column);
+        rLeftHandSideMatrix(row + NumNodes, column + NumNodes) = lhs_total(row, column);
+    }
+
+    // Applying wake condition on the AUXILIARY_VELOCITY_POTENTIAL dofs
+    if (data.distances[row] < 0.0)
+        for (unsigned int column = 0; column < NumNodes; ++column)
+            rLeftHandSideMatrix(row, column + NumNodes) = -lhs_total(row, column); // Side 1
+    else if (data.distances[row] > 0.0)
+        for (unsigned int column = 0; column < NumNodes; ++column)
+            rLeftHandSideMatrix(row + NumNodes, column) = -lhs_total(row, column); // Side 2
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::AssignRightHandSideWakeNode(
+    VectorType& rRightHandSideVector,
+    const BoundedVector<double, NumNodes>& rUpper_rhs,
+    const BoundedVector<double, NumNodes>& rLower_rhs,
+    const BoundedVector<double, NumNodes>& rWake_rhs,
+    const ElementalData<NumNodes, Dim>& rData,
+    unsigned int& rRow) const
+{
+    if (rData.distances[rRow] > 0.0){
+        rRightHandSideVector[rRow] = rUpper_rhs(rRow);
+        rRightHandSideVector[rRow + NumNodes] = rWake_rhs(rRow);
+    }
+    else{
+        rRightHandSideVector[rRow] = rWake_rhs(rRow);
+        rRightHandSideVector[rRow + NumNodes] = rLower_rhs(rRow);
+    }
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::ComputePotentialJump(const ProcessInfo& rCurrentProcessInfo)
+{
+    const array_1d<double, 3>& vinfinity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
+    const double vinfinity_norm = sqrt(inner_prod(vinfinity, vinfinity));
+
+    array_1d<double, NumNodes> distances;
+    GetWakeDistances(distances);
+
+    for (unsigned int i = 0; i < NumNodes; i++)
+    {
+        double aux_potential =
+            GetGeometry()[i].FastGetSolutionStepValue(AUXILIARY_VELOCITY_POTENTIAL);
+        double potential = GetGeometry()[i].FastGetSolutionStepValue(VELOCITY_POTENTIAL);
+        double potential_jump = aux_potential - potential;
+
+        if (distances[i] > 0)
+        {
+            GetGeometry()[i].SetValue(POTENTIAL_JUMP,
+                                      -2.0 / vinfinity_norm * (potential_jump));
+        }
+        else
+        {
+            GetGeometry()[i].SetValue(POTENTIAL_JUMP, 2.0 / vinfinity_norm * (potential_jump));
+        }
```

Comment:
```suggestion
        auto& r_node = GetGeometry()[i];
        const double aux_potential = r_node.FastGetSolutionStepValue(AUXILIARY_VELOCITY_POTENTIAL);
        const double potential = r_node.FastGetSolutionStepValue(VELOCITY_POTENTIAL);
        const double potential_jump = aux_potential - potential;
        const double aux = 2.0 / vinfinity_norm * (potential_jump));
        r_node.GetValue(POTENTIAL_JUMP) = (distances[i] > 0) ? - aux : aux;
```
Note that I'm using the `GetValue` method instead of the `SetValue`. The first returns a reference to the the variable value we want to change, so we can also use it to change the value. The point in here is that the `GetValue` is much faster than the `SetValue`.
As a rule of thumb, the `SetValue` should be only use once to do a first serial allocation of the memory.

Reply:
Quite nice! I like your suggestion because it is much more compact. Short question though, does this mean that we need to initialize the variable POTENTIAL_JUMP for all the elements at the beginning and that if we don't do this we might run into troubles when using GetValue in parallel? What is going to be GetValue's behavior if we don't initialize the variable? Seg fault? I remember having issues with this before. Actually I just checked and we changed from GetSolutionStepValue to SetValue because of this following the advice either of @jcotela or @msandre I don't remember.

Reply:
I need to write this down somewhere, I am explaining it every second day :smile: 
basically this: https://github.com/KratosMultiphysics/Kratos/pull/6691#issuecomment-609432609

Reply:
Exactly, so we are doing the first option because we want to allocate if it doesn't exist before, is this ok?

Reply:
Independent of whether you want to allocate or not I would say this is wrong since this is a node that can be accessed from multiple threads at the same time

Even more, this is an `=` and not a `+=` meaning whichever Elements node accesses this the last will put its value
This is basically the definition of a race-condition

I don't know what you want to do so I can't help